### PR TITLE
Fix: Refactor Waypoint Javascript Slot Machine

### LINKED ISSUE #4236: Javascript Slot Machine Reports "Not A Winner" When you Win
Challenge [Waypoint: Give your JavaScript Slot Machine some Stylish Images](http://www.freecodecamp.com/challenges/waypoint-give-your-javascript-slot-machine-some-stylish-images#?solution=fccss%0A%20%20function%20runSlots()%7B%0A%20%20%20%20var%20slotOne%3B%0A%20%20%20%20var%20slotTwo%3B%0A%20%20%20%20var%20slotThree%3B%0A%20%20%20%20%0A%20%20%20%20var%20images%20%3D%20%5B%22http%3A%2F%2Fi.imgur.com%2F9H17QFk.png%22%2C%20%22http%3A%2F%2Fi.imgur.com%2F9RmpXTy.png%22%2C%20%22http%3A%2F%2Fi.imgur.com%2FVJnmtt5.png%22%5D%3B%0A%20%20%20%20%0A%20%20%20%20slotOne%20%3D%20Math.floor(Math.random()%20*%20(3%20-%201%20%2B%201))%20%2B%201%3B%0A%20%20%20%20slotTwo%20%3D%20Math.floor(Math.random()%20*%20(3%20-%201%20%2B%201))%20%2B%201%3B%0A%20%20%20%20slotThree%20%3D%20Math.floor(Math.random()%20*%20(3%20-%201%20%2B%201))%20%2B%201%3B%0A%20%20%20%20%0A%20%20%20%20%24('.logger').html('')%3B%0A%20%20%20%20%24('.logger').html('Not%20A%20Win')%3B%0A%20%20%20%20%0A%20%20%20%20%2F%2F%20Only%20change%20code%20below%20this%20line.%0A%20%20%20%20%24(%24('.slot')%5B0%5D).html('%3Cimg%20src%20%3D%20%22'%20%2B%20images%5BslotOne-1%5D%20%2B%20'%22%3E')%3B%0A%20%20%20%20%24(%24('.slot')%5B1%5D).html('%3Cimg%20src%20%3D%20%22'%20%2B%20images%5BslotTwo-1%5D%20%2B%20'%22%3E')%3B%0A%20%20%20%20%24(%24('.slot')%5B2%5D).html('%3Cimg%20src%20%3D%20%22'%20%2B%20images%5BslotThree-1%5D%20%2B%20'%22%3E')%3B%0A%20%20%20%20%0A%20%20%20%20%0A%20%20%20%20%2F%2F%20Only%20change%20code%20above%20this%20line.%0A%20%20%20%20%0A%20%20%20%20if%20(slotOne%20%3D%3D%3D%20slotTwo%20%26%26%20slotTwo%20%3D%3D%3D%20slotThree)%20%7B%0A%20%20%20%20%20%20return%20slotOne%3B%0A%20%20%20%20%7D%0A%20%20%20%20%0A%20%20%20%20if(slotOne%20!%3D%3D%20undefined%20%26%26%20slotTwo%20!%3D%3D%20undefined%20%26%26%20slotThree%20!%3D%3D%20undefined)%7B%0A%20%20%20%20%20%20%24('.logger').html(slotOne)%3B%0A%20%20%20%20%20%20%24('.logger').append('%20'%20%2B%20slotTwo)%3B%0A%20%20%20%20%20%20%24('.logger').append('%20'%20%2B%20slotThree)%3B%0A%20%20%20%20%7D%0A%20%20%20%20%0A%20%20%20%20return%20%5BslotOne%2C%20slotTwo%2C%20slotThree%5D%3B%0A%20%20%7D%0A%0A%20%20%24(document).ready(function()%7B%0A%20%20%20%20%20%24('.go').click(function()%7B%0A%20%20%20%20%20%20%20runSlots()%3B%0A%20%20%20%20%20%7D)%3B%0A%20%20%20%7D)%3B%0Afcces%0A%20%0A%0A) has an issue.

User @bflatt72 noted that when you actually win it prints out "Not a Winner", rather than "Your are a Winner" or similar.  Challenge seed needs to be refactored.

My code:
```html
<script>
  function runSlots(){
    var slotOne;
    var slotTwo;
    var slotThree;
    
    var images = ["http://i.imgur.com/9H17QFk.png", "http://i.imgur.com/9RmpXTy.png", "http://i.imgur.com/VJnmtt5.png"];
    
    slotOne = Math.floor(Math.random() * (3 - 1 + 1)) + 1;
    slotTwo = Math.floor(Math.random() * (3 - 1 + 1)) + 1;
    slotThree = Math.floor(Math.random() * (3 - 1 + 1)) + 1;
    
    $('.logger').html('');
    $('.logger').html('Not A Win');
    
    // Only change code below this line.
    $($('.slot')[0]).html('<img src = "' + images[slotOne-1] + '">');
    $($('.slot')[1]).html('<img src = "' + images[slotTwo-1] + '">');
    $($('.slot')[2]).html('<img src = "' + images[slotThree-1] + '">');
    
    
    // Only change code above this line.
    
    if (slotOne === slotTwo && slotTwo === slotThree) {
      return slotOne; // Problem is probably here.
    }
    
    if(slotOne !== undefined && slotTwo !== undefined && slotThree !== undefined){
      $('.logger').html(slotOne);
      $('.logger').append(' ' + slotTwo);
      $('.logger').append(' ' + slotThree);
    }
    
    return [slotOne, slotTwo, slotThree];
  }

  $(document).ready(function(){
     $('.go').click(function(){
       runSlots();
     });
   });
</script>
 ```

### PR DESCRIPTION
**Pull request by aulisius**
Fixes #4236 

I refactored the code and moved the 'Not a winner' statement below.
-  When the numbers are equal then, it will print 'It's a Win' .
-  The statement 'Not a Win' will append to the .logger field instead of replacing the numbers.

Note: I haven't tested this locally. Haven't found the time to setup the dev env. Please check it locally.

### CODE REVIEW

**Comment by ltegman** [comment]
Woo -- the slot machine works! Only issue that because variations of this seed code are used in multiple lessons and the part you fixed is actually entered by the user in one of the earlier lessons you'll need to make a couple more changes. You'll need to alter the seed for the next lesson to match the changes you made and also change the lesson before so that the student is asked to enter in code that looks like what you've put in the seed.

**Comment by faizaanceg** [comment]
Is this ok? or should I add more in the description?

File: seed/challenges/basic-javascript.json [hunk 1]
```diff
@@ -1817,6 +1823,7 @@
         "    // Only change code above this line.",
         "    ",
         "    if (slotOne === slotTwo && slotTwo === slotThree) {",
+		"	 	$('.logger').html('It's A Win');",
         "      return slotOne;",
```
Comment (BerkeleyTrue) [inline seed/challenges/basic-javascript.json]: Change the quotes here to double quotes and make sure you have proper escaping.

Comment (BerkeleyTrue) [inline seed/challenges/basic-javascript.json]: Should be `$('.logger').html(\"It's A Win\");"

Comment (SaintPeter) [inline seed/challenges/basic-javascript.json]: There is a secondary issue that in the "spaces" before this line there are two tab characters which break the parser.  Those will need to be removed as well as the quotes being fixed.

**Comment by faizaanceg** [comment]
It still looks fine in my editor :/ so had to manually edit here. If this is fine, I'll squash the commits into one. @SaintPeter

**Comment by SaintPeter** [comment]
It's not a matter of how it looks, unless you can have "show whitespace" toggled on.  The issue is that there were two hard tabs (ASCII `#0009`) INSIDE the string.  Those are not allowed by the [JSON Spec](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf) (Page 10).  This was really tricky to track down.  So, thanks, I guess, for highlighting a deficiency in our linter.

There is still one lingering tab at the start of line 1826.  Go ahead an get rid of that and squash these down and I'll test again.

**Comment by faizaanceg** [comment]
Ah, ok. Didn't know about that :/, Sorry again for that! squashed the commits, hope this is fine now.
@SaintPeter

**Comment by SaintPeter** [comment]
@faizaanceg If it's any consolation, I didn't know either.  I've had a bit of "fun" learning about it today. ;)

**Comment by SaintPeter** [comment]
### "Waypoint: Add your JavaScript Slot Machine Slots" 
I believe the sample code is wrong and doesn't match the code in the next waypoint.
```js
if(slotOne !== slotTwo || slotTwo !== slotThree){
  return null;
}
```

It is note entirely clear to me where the "It's a Win" message should go in that waypoint, based on the instructions.

### For "Waypoint: Bring your JavaScript Slot Machine to Life"
This needs to be fixed:
```js
if (slotOne === slotTwo && slotTwo === slotThree) {
      $(".logger").html(" It's A Win")
      return slotOne;  // Should return null
    }
```
Same for the next waypoint.

### Waypoint: Give your JavaScript Slot Machine some Stylish Images
```js
$('.logger').append('Not A Win');
```
Needs a space before 'Not a Win'.

One last thing -
Maybe it should be "It's a Win!", because, yay, gambling.

**Comment by BerkeleyTrue** [comment]
@faizaanceg you'll need to do a rebase off staging for Travis tests to run.

**Comment by faizaanceg** [comment]
Done and done =). Playing around with Git is fun and annoying :P 

Is this ready to be merged @SaintPeter ?

**Comment by SaintPeter** [comment]
Ok, looking at "Waypoint: Add your JavaScript Slot Machine Slots":
> If they have, we should notify our user that they've won and we should return null.
>
> null is a JavaScript data structure that means nothing.
>
>The user wins when all the three numbers match. Let's create an if statement with multiple >conditions in order to check whether all numbers are equal.
>
>if(slotOne === slotTwo && slotTwo === slotThree){
>  return null;
>}
>
>Also, we need to show the user that he has won the game when he gets the same number in all the >slots.
>
>If all three numbers match, we should set the text "It's A Win" to the element with class logger or leave it as it is.

But the expected code is:
```js
if(slotOne === slotTwo || slotTwo === slotThree){
      
      $(".logger").html(" It's a Win");

      return null;
    }
```

I'm not sure if we should explicitly include the logger stuff in the sample code above or not.  We don't actually check for it in the tests.

Otherwise, this is ready to merge.

**Comment by faizaanceg** [comment]
@SaintPeter  If we add that line to the sample code, wouldn't we be giving away the *entire* answer? I thought the sample code should be a starting point, not a copy-paste answer.

**Comment by SaintPeter** [comment]
I'm not sure, actually, which is why I'm asking.  I'm just not sure if the instructions are enough:
>If all three numbers match, we should set the text "It's A Win" to the element with class logger or leave it as it is.

Maybe an alternate:
>If all three numbers match, we should also set the text "It's A Win" to the element with class logger.

**Comment by faizaanceg** [comment]
Hmm. That 'also' does make some difference. Hmm will do those changes

**Comment by SaintPeter** [comment]
This looks good @faizaanceg.
:+1: 

Thank you so much for taking the time to help get this right.  I know it's been a lot of work, but I think it will ultimately pay off with our users.

**Comment by ltegman** [comment]
Awesome job on this change both of you! :smile:

**Comment by faizaanceg** [comment]
:+1:

**Pull request merged**